spec/emerald/compiler_spec.rb

```diff
@@ -17,4 +17,24 @@
       expect(compiled_lisp).to eq("")
     end
   end
+
+  context "when the contents of the file define a variable" do
+    it "parses the lisp and returns an assigned global variable" do
+      lisp_file_contents = "(def y (- 1 6))"
+      compiled_lisp = Emerald::Compiler.new(lisp_file_contents).compile
+      expect(compiled_lisp).to eq("y = 1.0 - 6.0")
+    end
+
+    it "parses the lisp and returns an assigned local variable" do
+      lisp_file_contents = "(let [x 1 y (+ 3 4)] x)"
```

Comment:
So, I've ended up going for a Clojure-y approach and added in a new vector data structure here. It's not the approach I was planning to take but it ended up seeming like the cleanest way of parsing lists to solve the above issue. I tried separating the how lists were parsed depending on whether there were lists of equal level, like `(...)(...)`, and then based on matching parentheses, and also trying trying to do some graph theory type stuff, but they all started to get quite convoluted, and this ended up seeming to be the least messy...